crates/aranya-quic-channels/src/lib.rs

```diff
@@ -0,0 +1,293 @@
+    /// An infallible error.
+    #[error("infallible error: {0}")]
+    Infallible(#[from] std::convert::Infallible),
+    /// An io error.
+    #[error("io error: {0}")]
+    Io(#[from] std::io::Error),
+    /// An internal AQC error.
+    #[error("internal error: {0}")]
+    Internal(#[from] anyhow::Error),
+}
+
+/// Runs a server listening for quic channel requests from other peers.
+pub async fn run_channels(client: Arc<TMutex<AqcClient>>, mut server: Server) {
+    loop {
+        select! {
+            Some(conn) = server.accept() => {
+                match handle_connection(conn).await {
+                    Ok(Some(channel)) => {
+                        if client
+                            .lock()
+                            .await
+                            .new_channels
+                            .push(channel).is_err() {
+                            error!("Channel full. Unable to insert channel");
+                        }
+                    },
+                    Ok(None) => {
+                        // The connection was closed.
+                    },
+                    Err(e) => {
+                        error!(cause = ?e, "connection error");
+                    }
+                }
+            },
+        }
+    }
+}
+
+async fn handle_connection(mut conn: Connection) -> Result<Option<AqcChannel>> {
+    let stream = conn.accept_bidirectional_stream().await;
+    match stream {
+        Err(connection::Error::EndpointClosing { .. }) => Ok(None),
+        Err(e) => Err(e.into()),
+        Ok(None) => Ok(None),
+        Ok(Some(stream)) => handle_bidirectional_stream(conn, stream).await,
+    }
+}
+
+async fn handle_bidirectional_stream(
+    conn: Connection,
+    stream: BidirectionalStream,
+) -> Result<Option<AqcChannel>> {
+    let (recv, send) = stream.split();
+    // TODO: Use the SSL certificate to identify the channel.
+    let (channel, message_sender) = AqcChannel::new(send, recv);
+    tokio::spawn(handle_messages(conn, message_sender));
+
+    Ok(Some(channel))
+}
+
+async fn handle_messages(mut conn: Connection, sender: mpsc::Sender<Bytes>) {
+    while let Ok(Some(stream)) = conn.accept_receive_stream().await {
+        tokio::spawn(handle_message(stream, sender.clone()));
+    }
+}
+
+async fn handle_message(mut stream: ReceiveStream, sender: mpsc::Sender<Bytes>) {
+    let mut data = Vec::new();
+    loop {
+        match stream.receive().await {
+            Ok(Some(req)) => {
+                data.extend_from_slice(&req);
+            }
+            Ok(None) => {
+                if sender.send(Bytes::from(data)).await.is_err() {
+                    debug!("error sending to channel");
+                }
+                break;
+            }
+            Err(_) => {
+                debug!("error receiving from stream");
+                break;
+            }
+        }
+    }
+}
+
+#[derive(Eq, Hash, PartialEq, Clone, Copy, Debug)]
+/// Identifies a unique channel between two peers.
```

Comment:
We should differentiate from AFC in the docs:
```suggestion
/// Identifies a unique AQC channel between two peers.
```